### PR TITLE
Create a simple use example for a camera

### LINKED ISSUE #189: Set up example use case and document
@erinpougheon and @pbaudoz have an interest in this, so I will try to sketch out a game plan for this.

A minimal working example would constitute a testbed server that can run a camera service. This is then extendable arbitrarily to include more services. To make this work, we'd need:
1. A testbed configuration with `testbed.yml`
2. A CLI interface that allows the user to start the testbed server, `cli_interface.py`
3. A service configuration with a single service, a camera: `services.yml`
4. The simplest GUI viewer possible. Not sure if this should be a main window + camera viewer or just  a camera viewer, but I think how we launch the GUI right now does require that.

Numbers 1-3 should probably live in a new location in the project under `catkit2/examples/single_camera`, outside of the package. Number 4 should probably be the start of adding GUI elements to `catkit2` the package, and should go into `catkit2/catkit2/user_interface`. The concrete startup and setup of this would be called in the `cli_interface.py` of the example. 

In an ideal case, a second service will be added, a simulator service which uses the already existing `catkit2/simulator/simple_optical_model.py` - or we create an even simpler one with literally just a light source and a camera.

@ehpor @raphaelpclt any thoughts on this? I would in particular like some feedback on the scope of the example and the location of the individual files.

### PR DESCRIPTION
**Pull request by erinpougheon**
Fixes #189.

Also tried to update the link for the Allied Vision cameras in the ```environment.yml``` file to link it to Vimba X VmbPy library.

### CODE REVIEW

**Comment by erinpougheon** [comment]
After creating a new environment as described in the README of catkit (but with a new name):
The GUI is working in simulation. There is only a camera viewer.
image


I tested the code on hardware too, on the THD2 testbed, with the visitor_camera (called science camera in the testbed_example of catkit). It worked as expected after the installation of vmbpy for the allied vision camera.
Here are two screenshots of what it looks like, when the flat is on and without the flat, with the laser on.

image (14)
image (13)

File: catkit2/services/simple_simulator/simple_simulator.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def open(self):
 
         self.images = self.make_data_stream('images', 'float64', self.model.focal_grid.shape, 20)
 
-        self.update_atmosphere()
+        self.update_atmosphere
```
Comment (ivalaginja) [inline catkit2/services/simple_simulator/simple_simulator.py]: Was this change intentional? `self.atmosphere` is a callable, so it usually should keep the parentheses in order to be called properly.

File: catkit2/services/simple_simulator/simple_simulator.py [hunk 2]
```diff
@@ -51,6 +51,10 @@ def camera_readout(self, camera_name, power):
         except Exception as e:
             self.log.error(str(e))
 
+    def get_camera_power(self, camera_name):
+        wavefronts = self.model.get_wavefronts(camera_name)
+        return sum(wf.power for wf in wavefronts)
+
```
Comment (ivalaginja) [inline catkit2/services/simple_simulator/simple_simulator.py]: You'll need a second blank line here.

File: testbed_example/cli_interface.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+The control software for an example testbed.
+
+Usage:
+  catkit start server [--port PORT_ID] [--simulated] [--config_path PATH]...
+  catkit start gui [--port PORT_ID]
+  catkit (-h | --help)
+  catkit --version
+
+Options:
+  -p, --port PORT_ID         The port number on which the testbed server operates.
```
Comment (ivalaginja) [inline testbed_example/cli_interface.py]: The indentation is a little bit messed up here, suggestion:
```suggestion
  -p, --port PORT_ID       The port number on which the testbed server operates.
```

File: testbed_example/config/services.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
+simulator:
+  service_type: simple_simulator
+  requires_safety: false
```
Comment (ivalaginja) [inline testbed_example/config/services.yml]: Since we don't have a safety service, we might actually be able to delete this keyword. Could you remove `requires_safety` both here and in the science_camera, and check whether everything still works as expected?

**Comment by erinpougheon** [comment]
For the Vimba update in the ```environment.yml```, I let the previous link, because when I try to update it, I have a GitHub error, saying that it can be install from git (which in logical regarding the documentation of vmbpy of Vimba X). My suggestion might be to delete completly this line, but I'll wait for you opinion.

File: testbed_example/config/services.yml [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+  offset_x: 0
+  offset_y: 0
+  width: 400
+  height: 400
+  sensor_width: 1456
+  sensor_height: 1088
+  exposure_time: 200
+  gain: 0
+  flux: 100000000
```
Comment (ivalaginja) [inline testbed_example/config/services.yml]: Same for the flux, I am not sure whether the simple simulator/optical model actually uses it. Could you try running without it?

File: testbed_example/config/testbed.yml [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+startup_services: []
+base_data_path:
+  default: !path "~/example_data"
+support_data_path:
+  default: !path "~/example_support_data"
+
+base_experiment_path: "{simulator_or_hardware}/{date_and_time}_{experiment_name}/"
+sub_experiment_path: "{experiment_id:03d}_{experiment_name}/"
+
+long_term_monitoring_path: "{simulator_or_hardware}/long_term_monitoring/"
```
Comment (ivalaginja) [inline testbed_example/config/testbed.yml]: Please add a blank line at the end of this file.

File: testbed_example/config/testbed.yml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+default_port: 1234
+simulator:
+  service_id: simple_simulator
+startup_services: []
+base_data_path:
+  default: !path "~/example_data"
```
Comment (ivalaginja) [inline testbed_example/config/testbed.yml]: It might be better to rename this to "catkit2_example_data".

File: testbed_example/config/testbed.yml [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+support_data_path:
+  default: !path "~/example_support_data"
```
Comment (ivalaginja) [inline testbed_example/config/testbed.yml]: And then here "catkit2_example_support_data". A bit long, but shouldn't be a problem.

File: testbed_example/user_interface/assets/icon.jpg
Comment (ivalaginja) [inline testbed_example/user_interface/assets/icon.jpg:1]: I think this quick and dirty trial does not need to become any icon ^^ Can you keep the example without one for not?

Comment (ivalaginja) [reply testbed_example/user_interface/assets/icon.jpg:1]: Worst case, you could make a blank square in Inkscape and just write "catkit2" inside of it and use that.

File: testbed_example/user_interface/bench_display.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+            if not hasattr(self, 'open_' + info['opens']):
+                raise ValueError(f'Viewer {info["opens"]} is not known.')
+            button.clicked.connect(functools.partial(getattr(self, 'open_' + info['opens']), info['device_name']))
+
+            self.buttons.append(button)
+
+        self.make_context_menu()
+
+
+    def make_context_menu(self):
```
Comment (ivalaginja) [inline testbed_example/user_interface/bench_display.py]: Since this method is completely commented out, can you delete it? It might mean that you need to delete its usages, which is fine since it doesn't do anything.

Comment (ivalaginja) [reply testbed_example/user_interface/bench_display.py]: Or, delete most of the comments and leave just a single group of commands so that it presents an example, for example the poking action. We'd have to keep line 154 in this case too I think.

File: testbed_example/user_interface/bench_display.py [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+        scaling = min(self.width() / self.schematic_size.width(), self.height() / self.schematic_size.height())
+
+        # Set positions and sizes of buttons.
+        for button in self.buttons:
+            pos = button.relative_position
+            button.move(pos[0] * self.schematic_size.width() * scaling, pos[1] * self.schematic_size.height() * scaling)
+            button.resize(scaling * 70, scaling * 70)
+
+
+
```
Comment (ivalaginja) [inline testbed_example/user_interface/bench_display.py]: This looks like two blank lines too many.

File: testbed_example/user_interface/camera_viewer.py [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+        vbox_sidebar.addWidget(w)
+
+        # Add parameters access
+        self.p = pg.parametertree.Parameter.create(name='Camera parameters', type='group', children=params)
+        self.tree = pg.parametertree.ParameterTree(self.sidebar)
+        self.tree.setParameters(self.p, showTop=False)
+        vbox_sidebar.addWidget(self.tree)
+
+        # Set to values of camera
+        self.p.param('Exposure', 'ExposureTime').setValue(self.camera.exposure_time / 1e6)
```
Comment (ivalaginja) [inline testbed_example/user_interface/camera_viewer.py]: This is something I want to address, although maybe not in this PR. This is a hard-coded conversion factor for the units of the camera exposure time, which is not fantastic in the example. Could you open an issue that says: "Remove hard-coded exposure time scaling in simple camera viewer" and link it here?

File: testbed_example/user_interface/common.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+
+    @QtCore.pyqtProperty(float)
+    def handle_position(self):
+        return self._handle_position
+
+    @handle_position.setter
+    def handle_position(self, pos):
+        self._handle_position = pos
+        self.update()
+
```
Comment (ivalaginja) [inline testbed_example/user_interface/common.py]: Let's add another blank line here.

File: testbed_example/user_interface/logging_display.py [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        self.listen_timer.setInterval(100)
+        self.listen_timer.start()
+
+    def clear_filter(self):
+        self.search_filter.setText('')
+
+        i = self.threshold_dropdown.findText('Info')
+        if i >= 0:
+            self.threshold_dropdown.setCurrentIndex(i)
+
```
Comment (ivalaginja) [inline testbed_example/user_interface/logging_display.py]: Another blank line.

**Review by ivalaginja — COMMENTED** [review]
Looking good! Just some small stuff to address.

The big question we need to answer here is whether we can move some of the GUI stuff into the actual `catkit2` package. @erinpougheon how about you address the comments I left, and then we try to estimate together whether this is doable?

File: catkit2/services/simple_simulator/simple_simulator.py (hunk 1 above)
Comment (erinpougheon) [reply catkit2/services/simple_simulator/simple_simulator.py]: Yes it is. Before deleting the parenthesis, I had an error ```TypeError: update_atmosphere() takes 1 positional argument but 2 were given```. It is the only solution I found to solve this problem.

**Comment by ivalaginja** [comment]
> For the Vimba update in the `environment.yml`, I let the previous link, because when I try to update it, I have a GitHub error, saying that it can be install from git (which in logical regarding the documentation of vmbpy of Vimba X). My suggestion might be to delete completly this line, but I'll wait for you opinion.

@erinpougheon ok. Can you open an issue for this ("Update Vimba Python package installation to VimbaX" or something like that) so that we don't forget about it?

File: testbed_example/config/services.yml (hunk 5 above)
Comment (erinpougheon) [reply testbed_example/config/services.yml]: image

After removing the safety, I have this error, so I put the lines for 'require safety' back in the config file.
The 'flux' can be removed without any problem.

File: testbed_example/user_interface/camera_viewer.py (hunk 11 above)
Comment (erinpougheon) [reply testbed_example/user_interface/camera_viewer.py]: Here is the corresponding new issue :  https://github.com/spacetelescope/catkit2/issues/228

File: testbed_example/user_interface/bench_display.py (hunk 9 above)
Comment (erinpougheon) [reply testbed_example/user_interface/bench_display.py]: I let the poking action example, and deleted the rest.

**Comment by erinpougheon** [comment]
> > For the Vimba update in the `environment.yml`, I let the previous link, because when I try to update it, I have a GitHub error, saying that it can be install from git (which in logical regarding the documentation of vmbpy of Vimba X). My suggestion might be to delete completly this line, but I'll wait for you opinion.
> 
> @erinpougheon ok. Can you open an issue for this ("Update Vimba Python package installation to VimbaX" or something like that) so that we don't forget about it?

Done, here is the new issue : https://github.com/spacetelescope/catkit2/issues/229

**Comment by erinpougheon** [comment]
I changed everything you requested !

File: testbed_example/user_interface/bench_display.py (hunk 9 above)
Comment (ivalaginja) [reply testbed_example/user_interface/bench_display.py]: Could you add a blank line after the `# self.setContextMenuPolicy(QtCore.Qt.ActionsContextMenu)` to make it obvious that it's a general command?

File: catkit2/services/simple_simulator/simple_simulator.py (hunk 1 above)
Comment (ivalaginja) [reply catkit2/services/simple_simulator/simple_simulator.py]: I find that weird. Let me have a look at this and I'll come back to you, because currently I can't find an explanation for this behavior.

**Review by ivalaginja — CHANGES_REQUESTED** [review]
One mini thing to fix, please, and the other comment is more a note to myself - consider this PR done from your side, and I will try to check why I find that one callable call weird.

**Comment by ivalaginja** [comment]
While trying to figure out what is happening with the `update_atmosphere()` method, I saw that the input to adding a callback function is not only the wrong way around where you fixed it, but also in the DM method:
https://github.com/spacetelescope/catkit2/blob/5c2f7a9e2e2b665577a03b2bf1d1ed8c7e5d1316/catkit2/services/simple_simulator/simple_simulator.py#L30

Could you fix that one as well please? It is not used in the example you set up so we didn't notice, but since we're already on it, let's fix that one too.

File: catkit2/services/simple_simulator/simple_simulator.py (hunk 1 above)
Comment (ivalaginja) [reply catkit2/services/simple_simulator/simple_simulator.py]: @ehpor any clue what might be happening here? @erinpougheon could you post the error message you're getting for this here please?

Comment (erinpougheon) [reply catkit2/services/simple_simulator/simple_simulator.py]: image (13)

Comment (raphaelpclt) [reply catkit2/services/simple_simulator/simple_simulator.py]: Hmmm I think you're avoiding the error, but I don't think the atmosphere works either. My intuition is that it just doesn't do anything. 

If you look at the simulator.py code here, https://github.com/spacetelescope/catkit2/blob/a7c17df19636676151d965e67d0085367ba14769/catkit2/simulator/simulator.py#L131C13-L131C41, it calls the callback function with time as an argument. So it will call ``self.update_atmosphere(t)``, that has 2 arguments (self, and t). You would need to define ``def update_atmosphere(self, t)`` or something like that.

Also, looking at the code, I feel the ``simple_simulator`` on develop is bugged.

**Comment by erinpougheon** [comment]
@ivalaginja changed the simulator and optical model used in the example, so we don't have the same problem with the ```update_atmosphere()```. (I think we can maybe open a new issue to just adress this simulator problem, I don't know).

Now, the example is working in simulation (evn though there is no light on it, because light is not implmented in the new example of simulation). Here is the result on the simulated server:
image

Hardware test on the THD2 testbed: Everything works as expected. I was able to have the viewer (we can see things because I turned on a laser). I could change the ROI, the gain, the exposure time etc.
image (14)


I think this PR is ready for final review and can be merged. I added the parenthesis back in the update_atmosphere function, in case we want it to be a separate issue.

File: testbed_example/example_optical_model.py [hunk 14]
```diff
@@ -30,6 +30,10 @@ def pupil_grid(self):
 
         return hcipy.make_uniform_grid(dims, size)
 
+    @property
+    def focal_grid(self):
+        return hcipy.make_focal_grid(12, 32)
+
```
Comment (ivalaginja) [inline testbed_example/example_optical_model.py]: Do you think wherever you needed the `focal_grid` you could just use the `detector_grid` below? I would like the grid parameters to be read from the configuration, and I was trying to call things in the detector plane (where the focal grid is) with "detector" to keep things more or less consistent.

File: testbed_example/services/example_simulator/example_simulator.py [hunk 15]
```diff
@@ -12,7 +12,7 @@ def __init__(self):
         super().__init__('example_simulator')
 
     def open(self):
-        self.model = ExampleOpticalModel()
+        self.model = ExampleOpticalModel(self.testbed.config)
```
Comment (ivalaginja) [inline testbed_example/services/example_simulator/example_simulator.py]: Nice catch on this - I also think it needs to be `self.testbed.config['simulator']`? Otherwise I am not sure what pool of configurations it tries to read, between all the different configuration files the example has. It might be automatically choosing the right one, but it's better to be explicit.

Comment (erinpougheon) [reply testbed_example/services/example_simulator/example_simulator.py]: I made it go on the simulator file in the optical model, but I can defenitly change that you're right.

**Comment by erinpougheon** [comment]
image

Here is the viewer (in simulation, because we changed nothing that has to do with the hardware) after the changes, everything works.

**Review by ivalaginja — DISMISSED** [review]
Good by me. Please wait with merging until we hear back from @raphaelpclt.

File: testbed_example/cli_interface.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+Usage:
+  catkit start server [--port PORT_ID] [--simulated] [--config_path PATH]...
+  catkit start gui [--port PORT_ID]
+  catkit (-h | --help)
+  catkit --version
+
+Options:
+  -p, --port PORT_ID      The port number on which the testbed server operates.
+                          Defaults to the port specified in the config file.
+  --simulated             Whether the testbed server should be run in simulated mode or not.
```
Comment (raphaelpclt) [inline testbed_example/cli_interface.py]: I am not sure - shouldn't ``--simulated`` be the default option or hardcoded? This example will never be running not simulated I guess.

File: testbed_example/config/testbed.yml [hunk 17]
```diff
@@ -0,0 +1,15 @@
...
+startup_services: []
+base_data_path:
+  default: !path "~/catkit2_example_data"
+support_data_path:
+  default: !path "~/catkit2_example_support_data"
+
+base_experiment_path: "{simulator_or_hardware}/{date_and_time}_{experiment_name}/"
+sub_experiment_path: "{experiment_id:03d}_{experiment_name}/"
+
+long_term_monitoring_path: "{simulator_or_hardware}/long_term_monitoring/"
```
Comment (raphaelpclt) [inline testbed_example/config/testbed.yml:15]: These path are not relevant in this example

**Review by raphaelpclt — COMMENTED** [review]
The code looks good overall. A few things can be removed (like experiment paths in the config - catkit2 doesn't implement experiment structures). Still pending discussion on what to let public or not.
